### PR TITLE
Improve phpstan configuration

### PR DESCRIPTION
**Pull request by Landerstraeten**
| Q             | A
| ------------- | ---
| Branch        | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Documented?   | yes
| Fixed tickets | #400

### CODE REVIEW

File: doc/tasks/phpstan.md [hunk 1]
```diff
@@ -38,6 +40,18 @@ With this parameter you can specify the path your project's configuration file.
 
 With this parameter you can set the level of rule options - the higher the stricter.
 
+**ignore_patterns**
+
+*Default: []*
+
+This is a list of patterns that will be forced for analysis even when the file or path is ignored.
+
+**ignore_patterns**
```
Comment (Kanti) [inline doc/tasks/phpstan.md]: ignore_patterns is duplicated here.

Comment (Landerstraeten) [reply doc/tasks/phpstan.md]: Wups. Will fix the typo

File: src/Task/PhpStan.php [hunk 2]
```diff
@@ -56,7 +60,15 @@ public function canRunInContext(ContextInterface $context)
...
         $config = $this->getConfiguration();
-        $files = $context->getFiles()->extensions($config['triggered_by']);
+        $files = $context
+            ->getFiles()
+            ->notPaths($config['ignore_patterns'])
+            ->extensions($config['triggered_by']);
+
+        $forcedFiles = $context->getFiles()->paths($config['force_patterns']);
+        foreach ($forcedFiles as $forcedFile) {
+            $files->add($forcedFile);
```
Comment (veewee) [inline src/Task/PhpStan.php]: Shouldn't we do some intersection here to make sure the files aren't added twice?

File: src/Collection/FilesCollection.php [hunk 3]
```diff
@@ -104,6 +104,24 @@ public function notPath($pattern)
...
+     *
+     * You can use patterns (delimited with / sign) or simple strings.
+     *
+     * $collection->notPaths(['/^spec\/','/^src\/'])
+     *
+     * @param array $pattern
+     *
+     * @return FilesCollection
+     */
+    public function notPaths(array $pattern)
```
Comment (veewee) [inline src/Collection/FilesCollection.php]: Can you change `notPath()` to use this method (similar to `path()` and `paths()`) ?

File: src/Task/PhpStan.php [hunk 4]
```diff
@@ -56,7 +60,17 @@ public function canRunInContext(ContextInterface $context)
     public function run(ContextInterface $context)
     {
         $config = $this->getConfiguration();
-        $files = $context->getFiles()->extensions($config['triggered_by']);
+        $files = $context
+            ->getFiles()
+            ->notPaths($config['ignore_patterns'])
+            ->extensions($config['triggered_by']);
+
+        $forcedFiles = $context->getFiles()->paths($config['force_patterns']);
```
Comment (veewee) [inline src/Task/PhpStan.php]: Can we move these lines that are forcing the files to be available to the FilesCollection?
This way we can do something like:
`$files->ensureFiles($context->getFiles()->paths($config['force_patterns]))`

File: src/Task/PhpStan.php [hunk 5]
```diff
@@ -56,7 +60,17 @@ public function canRunInContext(ContextInterface $context)
     public function run(ContextInterface $context)
     {
         $config = $this->getConfiguration();
-        $files = $context->getFiles()->extensions($config['triggered_by']);
+        $files = $context
```
Comment (veewee) [inline src/Task/PhpStan.php]: Can you add tests for both `ignore_patterns` and `force_patterns` to make sure the tasks behaves as it should with those 2 additional parameters?

**Comment by aloupfor** [comment]
Thanks for the PR.
I opened an issue regarding phpstan configuration so I am happy to see that this was done.

**Pull request merged**